readthedocs/doc_builder/environments.py

```diff
@@ -267,7 +270,88 @@ def get_wrapped_command(self):
                                    for part in self.command]))))
 
 
-class BuildEnvironment(object):
+class BaseEnvironment(object):
+
+    """
+    Base environment class.
+
+    Used to run arbitrary commands outside a build.
+    """
+
+    def __init__(self, project, environment=None):
+        # TODO: maybe we can remove this Project dependency also
+        self.project = project
+        self.environment = environment or {}
+        self.commands = []
+
+    def record_command(self, command):
+        pass
+
+    def _log_warn_only(self, msg):
+        log.warn(LOG_TEMPLATE.format(
+            project=self.project.slug,
+            version='latest',
+            msg=msg,
+        ))
+
+    def run(self, *cmd, **kwargs):
+        """Shortcut to run command from environment."""
+        return self.run_command_class(cls=self.command_class, cmd=cmd, **kwargs)
+
+    def run_command_class(self, cls, cmd, warn_only=False, **kwargs):
+        """
+        Run command from this environment.
+
+        Use ``cls`` to instantiate a command
+
+        :param warn_only: Don't raise an exception on command failure
+        """
+        # Remove PATH from env, and set it to bin_path if it isn't passed in
+        env_path = self.environment.pop('BIN_PATH', None)
+        if 'bin_path' not in kwargs and env_path:
+            kwargs['bin_path'] = env_path
+        assert 'environment' not in kwargs, "environment can't be passed in via commands."
+        kwargs['environment'] = self.environment
+
+        # TODO: ``build_env`` is passed as ``kwargs`` when it's called from a ``*BuildEnvironment``
+        # kwargs['build_env'] = self
+
+        build_cmd = cls(cmd, **kwargs)
+        self.commands.append(build_cmd)
+        build_cmd.run()
+
+        # TODO: I don't like how it's handled this entry point here
```

Comment:
This TODO comment was moved accidentally. It belongs to this line: https://github.com/rtfd/readthedocs.org/pull/3520/files/7f6c098983acb4bb3afc90e5c780af7dae9ec9d8#r163242699

I agree with the return states, but I'd say that the last two shouldn't exist at all. The rest, are needed.

Reply:
> Command failed, but report as pass? (I don't know if this still exists though)

This is tricky, but it does exist. Although, we are not using it but maybe we need it.

For example, `hg tags` command. I suppose that we want to `record=True` it but we want to `warn_only=True` also, since it could fail but we don't want the Build to fail. Besides, if it's recorded and failed (exit status != 0) it will be presented in _red_ to the user and we don't want that.

Is that a good example of this use case?

Reply:
Sumarizing, I think we want the `warn_only` commands to be recorded but don't interfers in the build status nor in the command itself.

Maybe what we need to change is the `BuildCommandResultMixin.successful` property to something like:

    return self.exit_code == 0 and self.warn_only == False

but that would involve to save the `warn_only` in the database also and I don't know if this is getting more and more complicated :/

Reply:
Another possible idea is that if `warn_only` we record it, but we set the `exit_code` as `0` (this could end up with other problems like the Latex build fail and we will never know it, even when opening the Build interface since it won't be marked as failed -in red).

So, add another attribute? `warn_only` _and_ `force_success` 😛 ?

Reply:
👍  Its sounding like we also want a `force_success` option as well here.

Reply:
Meh, I just realized that we need something like `record_as_success` instead of `force_success` because we are using `git status` exit code to decide if `clone` or `pull`, so we _need_ the real exit code in the flow but we want to save a _fake_ one in the databse...